nep-monitor.mediawiki

```diff
@@ -0,0 +1,171 @@
+[[Image:https://raw.githubusercontent.com/ixje/proposals/monitor/assets/nep-monitor/handshake.png]]
+
+'''Implementation caveats: '''
+
+Separate monitoring clients from regular node clients into their own connection pool to avoid maximum node connection limitations being hit due to connected monitoring clients.
+
+===Monitor payload===
+The payload data is a serialized JSON object containing the attributes listed in the next section. An example using neo-python for returning a monitor network message with 2 attributes for the MainNet would look as follows:
+<pre>
+json_data = {"Description": "my node", "Agent": "neopython-0.8.2"}
+m = Message(magic=0x00746e41, command='monitor', payload=MonitorPayload(data=json.dumps(json_data)))
+data_bytearray = Helper.ToArray(m)
+</pre>
+
+===Data attributes===
+The following data attributes shall be encapsulated in the <code>monitor</code> payload.
+
+{| class="wikitable"
+! Attribute/Key
+! Return type/value
+! Description
+|-
+| Description
+| uint8[20]
+| User friendly identification string
+|-
+| Agent
+| uint8[20]
+| Client identification string <code><name>-<version> </code>(i.e. "neopython-0.8.2")
+|-
+| BlockHeight
+| uint32
+| Index of the last block in the chain
+|-
+| BlockHash
+| uint160
+| Hash value of the last block in the chain
+|-
+| BlockTxCount
+| uint32
+| Number of transactions in the last block
+|-
+| BlockSize
+| uint32
+| The number of bytes the last block's data consist off<br/>
+Note: need to specify what exact data to count
+|-
+| BlockReceivedTime
+| uint32
+| Time in seconds from Unix EPOCH since receiving the latest block from the network
+|-
+| PeerCount
+| uint16
+| The combined number of peers connected on the P2P and/or websocket
+|-
+| InterfaceState
+| uint8
+| A flag indicating available interfaces<br/>
+0x1: P2P<br/>
+0x2: Websocket<br/>
+0x4: RPC<br/>
+0x8: REST<br/>
+0x10: reserved<br/>
+0x20: reserved
+|-
+| UnconfirmedTxCount
+| uint32
+| The number of transactions in the mempool after committing the block
+|-
+| IncomingTxPerSecond
+| uint32
+| The number of transactions submitted to the mempool between the LastBlock and LastBlock-1
+|-
+| ProcessedTxPerSecond
+| Fixed8
+| The number of transactions processed per second in the last block<br/>
+Calculated as: number of transactions in block divided by the total time in seconds needed to process the transactions
+|-
+| Uptime
+| uint32
+| Time in seconds from Unix EPOCH since the node started running without interruption.
+|}
+
+==Rationale==
+
+===Monitor interface===
+'''Shared vs separate interface'''
+
+By establishing a separate channel on the regular P2P/WebSocket port, as opposed to creating the service on its own port, we avoid losing the ability to collect statistics whenever a node is actively participating in the network. 
```

Comment:
Splitting channels by interface or port works well for monitoring traffic using classical network analysis tools.

Each node will naturally be capable of collecting stats, but it isn't clear that these should be managed and delivered on the same node that is computing chain. The data could be offloaded to a monitoring node that is essentially a monitoring data mirror that only computes monitoring client requests and delivery. Each node would have a monitor mirror. This adds more nodes to the network, but helps with scalability and lightens node overhead.

Reply:
Changes to existing "chain" node code to facilitate the mirror-node monitoring approach could be as simple as additional logging calls. The mirror could just consume logs using standard log management tools, for example. Optimal performance must be observed during selection.

Reply:
In the case of persistent stat collection, database integration is heavier performance-wise as the chain node has to handle queries, but may ultimately be a requirement for the types of data being requested. It does lend some consideration to the notion of chain node optimization via off-box database coupling via a logging mechanism,  just spitballing here.

Reply:
We're aware of an easier initial implementation when the interface is on a separate port, but at the same time it means more conscious work for the node hoster to eventually set it up. Obviously I'm biased here, but I rather have a higher upfront implementation cost and then later have a low barrier for anyone hosting a node to share their data, than vice versa. 

I don't see much benefit of a monitoring node, unless you support custom request profiles. The full data set still has to be collected on the _chain node_ and send to this monitoring node. While we can't foresee it, I don't think we'll see such a huge number of data collection clients that off-loading becomes necessary. If that eventually does become the case, then you can always add that later by adding a client filter on the monitor channel that only accepts the off loading node and the rest can make use of that. 

Logging calls are heavy and disk I/O bound, which I believe will bring down the node performance as now the disk is shared with LevelDB access for committing blocks. The current proposal allows for all data to stay in memory (the static parts like `Description` and `Agent` can be cached, the rest only needs a 1 time snapshot that gets overwritten each round)

> In the case of persistent stat collection, database integration is heavier performance-wise as the chain node has to handle queries, but may ultimately be a requirement for the types of data being requested.
Can you give an example of an attribute that requires this? We tried to write it as such that no historical data has to be collected for longer than the block interval. This is because we believe it is up to statistic websites (i.e. HappyNodes) to track this.